[CLOUDSTACK-10323] Allow changing disk offering during volume migration

rafaelweingartner opened the pull request:
This is a continuation of work developed on PR #2425 (CLOUDSTACK-10240), which provided root admins an override mechanism to move volumes between storage systems types (local/shared) even when the disk offering would not allow such operation. To complete the work, we will now provide a way for administrators to enter a new disk offering that can reflect the new placement of the volume. We will add an extra parameter to allow the root admin inform a new disk offering for the volume. Therefore, when the volume is being migrated, it will be possible to replace the disk offering to reflect the new placement of the volume.

The API method will have the following parameters:

* storageid (required)
* volumeid (required)
* livemigrate(optional)
* newdiskofferingid (optional) – this is the new parameter

The expected behavior is the following:

* If “newdiskofferingid” is not provided the current behavior is maintained. Override mechanism will also keep working as we have seen so far.
* If the “newdiskofferingid” is provided by the admin, we will execute the following checks
    * new disk offering mode (local/shared) must match the target storage mode. If it does not match, an exception will be thrown and the operator will receive a message indicating the problem.
    * we will check if the new disk offering tags match the target storage tags. If it does not match, an exception will be thrown and the operator will receive a message indicating the problem.
    * check if the target storage has the capacity for the new volume. If it does not have enough space, then an exception is thrown and the operator will receive a message indicating the problem.
    * check if the size of the volume is the same as the size of the new disk offering. If it is not the same, we will ALLOW the change of the service offering, and a warning message will be logged.

We execute the change of the Disk offering as soon as the migration of the volume finishes. Therefore, if an error happens during the migration and the volume remains in the original storage system, the disk offering will keep reflecting this situation.

Attached is a screenshot to show how the migrate option will look like:
![changeintheui](https://user-images.githubusercontent.com/4129005/37377238-729855d2-2707-11e8-8870-9c50ba614b20.png)

borisstoyanov commented:
thanks for this PR @rafaelweingartner, are you looking into adding a marvin tests for this ?
I'll wait on your response and kick the smoketests.

rafaelweingartner commented:
I added unit test for the code I changed. I was not planning to create a Marvin tests though.
If you guys think it is a requirement, I can look into it.

borisstoyanov commented:
@blueorangutan package

borisstoyanov commented:
ok @rafaelweingartner, let me run some tests

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1777

borisstoyanov commented:
@blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-2363)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 31355 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2486-t2363-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_public_ip_range.py
Intermitten failure detected: /marvin/tests/smoke/test_ssvm.py
Intermitten failure detected: /marvin/tests/smoke/test_templates.py
Intermitten failure detected: /marvin/tests/smoke/test_usage.py
Intermitten failure detected: /marvin/tests/smoke/test_volumes.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Intermitten failure detected: /marvin/tests/smoke/test_hostha_kvm.py
Smoke tests completed. 62 look OK, 5 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_04_extract_template | `Failure` | 128.33 | test_templates.py
ContextSuite context=TestISOUsage>:setup | `Error` | 0.00 | test_usage.py
test_06_download_detached_volume | `Failure` | 137.59 | test_volumes.py
test_01_redundant_vpc_site2site_vpn | `Failure` | 325.32 | test_vpc_vpn.py
test_hostha_enable_ha_when_host_in_maintenance | `Error` | 0.45 | test_hostha_kvm.py

borisstoyanov commented:
@rafaelweingartner I've created a marvin test but I'm not able to push to your repo. How can I contribute?

rafaelweingartner commented:
Thanks!
You can open a PR against my branch, then I can merge it here.

borisstoyanov commented:
I'm not sure I'm doing this right.. but git says there's nothing to compare.. 
![screen shot 2018-03-15 at 12 51 51](https://user-images.githubusercontent.com/13551960/37459115-c621161e-284f-11e8-8e42-c28590568c96.png)
any idea? @rafaelweingartner

rafaelweingartner commented:
Did you pushed your changes to your branch in the remote repository?
I am checking your branch, but there is no new commit there.

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✖debian. JID-1781

borisstoyanov commented:
@blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

DaanHoogland commented:
debian packaging failed and so did jenkins and travis!

rafaelweingartner commented:
Yes. However, that is a consequence of the last merge and is fixed with: #2491

borisstoyanov commented:
@blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

borisstoyanov commented:
@blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rafaelweingartner commented:
@DaanHoogland now everything is ok again.

DaanHoogland commented:
thanks @rafaelweingartner , didn't notice the relation

nitin-maharana commented on server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:
```diff
@@ -2097,26 +2079,102 @@ public Volume migrateVolume(MigrateVolumeCmd cmd) {
+    /**
+     * Retrieve the new disk offering UUID that might be sent to replace the current one in the volume being migrated.
+     * If no disk offering UUID is provided we return null. Otherwise, we perform the following checks.
+     * <ul>
```
@rafaelweingartner, I think you have added few HTML tags by mistake here.

rafaelweingartner replied:
It is not by mistake. It is on purpose to format the java documentation. Using them you are able to create javadocs as this one:
![javadoc](https://user-images.githubusercontent.com/4129005/37590952-a3137f44-2b48-11e8-9408-a32906dc0732.png)

nitin-maharana replied:
Got it. Thanks.

nitin-maharana commented on server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:
```diff
@@ -2097,26 +2079,102 @@ public Volume migrateVolume(MigrateVolumeCmd cmd) {
+    /**
+     * Performs the validations required for replacing the disk offering while migrating the volume of storage. If no new disk offering is provided, we do not execute any validation.
+     * If a disk offering is informed, we then proceed with the following checks.
+     * <ul>
```
Same here.

rafaelweingartner commented:
@blueorangutan package

blueorangutan commented:
@rafaelweingartner a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1790

borisstoyanov commented:
@blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-2393)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 32550 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2486-t2393-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_certauthority_root.py
Intermitten failure detected: /marvin/tests/smoke/test_deploy_virtio_scsi_vm.py
Intermitten failure detected: /marvin/tests/smoke/test_primary_storage.py
Intermitten failure detected: /marvin/tests/smoke/test_public_ip_range.py
Intermitten failure detected: /marvin/tests/smoke/test_snapshots.py
Intermitten failure detected: /marvin/tests/smoke/test_templates.py
Intermitten failure detected: /marvin/tests/smoke/test_usage.py
Intermitten failure detected: /marvin/tests/smoke/test_vm_life_cycle.py
Intermitten failure detected: /marvin/tests/smoke/test_volumes.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermitten failure detected: /marvin/tests/smoke/test_host_maintenance.py
Intermitten failure detected: /marvin/tests/smoke/test_hostha_kvm.py
Smoke tests completed. 58 look OK, 9 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_add_primary_storage_disabled_host | `Error` | 0.71 | test_primary_storage.py
test_01_primary_storage_nfs | `Error` | 0.11 | test_primary_storage.py
ContextSuite context=TestStorageTags>:setup | `Error` | 0.19 | test_primary_storage.py
test_02_list_snapshots_with_removed_data_store | `Error` | 1.15 | test_snapshots.py
test_04_extract_template | `Failure` | 128.28 | test_templates.py
ContextSuite context=TestISOUsage>:setup | `Error` | 0.00 | test_usage.py
test_08_migrate_vm | `Error` | 15.71 | test_vm_life_cycle.py
test_06_download_detached_volume | `Failure` | 137.56 | test_volumes.py
test_04_rvpc_network_garbage_collector_nics | `Failure` | 448.41 | test_vpc_redundant.py
test_01_cancel_host_maintenace_with_no_migration_jobs | `Failure` | 0.09 | test_host_maintenance.py
test_02_cancel_host_maintenace_with_migration_jobs | `Error` | 2.23 | test_host_maintenance.py
test_hostha_enable_ha_when_host_in_maintenance | `Error` | 2.42 | test_hostha_kvm.py

borisstoyanov commented:
@blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-2396)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 32771 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2486-t2396-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_certauthority_root.py
Intermitten failure detected: /marvin/tests/smoke/test_volumes.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermitten failure detected: /marvin/tests/smoke/test_host_maintenance.py
Intermitten failure detected: /marvin/tests/smoke/test_hostha_kvm.py
Smoke tests completed. 64 look OK, 3 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_11_migrate_volume_and_change_offering | `Error` | 131.28 | test_volumes.py
test_04_rvpc_network_garbage_collector_nics | `Failure` | 537.61 | test_vpc_redundant.py
test_hostha_enable_ha_when_host_in_maintenance | `Error` | 2.96 | test_hostha_kvm.py
test_hostha_kvm_host_fencing | `Error` | 7.97 | test_hostha_kvm.py

rafaelweingartner commented:
@borisstoyanov are these errors something that I should take a look into? I noticed that the test you included is failing as well.

I look at the error of the new test, and it seems you are trying to change the service offering to a new one that does not have the tags of the target storage.

`Target Storage [id=0818cecf-5f99-3b1e-beb2-5f0669c7b5ba] tags [storage.overprovisioning.factor] does not match new disk offering [id=a2992784-662b-4cad-b833-38c46136f97e] tags [null].'}, accountid : u'9f16aecc-2c1d-11e8-8460-06d348010705'}\n"]`

borisstoyanov commented:
@rafaelweingartner let me run this again, don't know why it appeared.
@blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-2403)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 42229 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2486-t2403-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_affinity_groups_projects.py
Intermitten failure detected: /marvin/tests/smoke/test_affinity_groups.py
Intermitten failure detected: /marvin/tests/smoke/test_certauthority_root.py
Intermitten failure detected: /marvin/tests/smoke/test_deploy_virtio_scsi_vm.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_volumes.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermitten failure detected: /marvin/tests/smoke/test_hostha_kvm.py
Smoke tests completed. 61 look OK, 6 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_DeployVmAntiAffinityGroup_in_project | `Error` | 58.44 | test_affinity_groups_projects.py
test_DeployVmAntiAffinityGroup | `Error` | 63.06 | test_affinity_groups.py
test_04_rvpc_privategw_static_routes | `Failure` | 370.47 | test_privategw_acl.py
test_11_migrate_volume_and_change_offering | `Error` | 130.68 | test_volumes.py
test_04_rvpc_network_garbage_collector_nics | `Failure` | 540.19 | test_vpc_redundant.py
test_hostha_enable_ha_when_host_in_maintenance | `Error` | 2.44 | test_hostha_kvm.py
test_hostha_kvm_host_fencing | `Failure` | 675.83 | test_hostha_kvm.py
test_hostha_kvm_host_recovering | `Failure` | 623.78 | test_hostha_kvm.py

rafaelweingartner commented:
Now the errors changed, it increased the number of them. However, the new test case that was added keeps failing.

borisstoyanov commented:
@rafaelweingartner if we add this in the marvin tests it should clean before migration, can you just add it before it tries to migrate?
```
        if hasattr(pool, 'tags'):
            StoragePool.update(self.apiclient, id=pool.id, tags="")
```

rafaelweingartner commented:
@borisstoyanov I ammended the code as you suggested.

rafaelweingartner commented:
@blueorangutan package

blueorangutan commented:
@rafaelweingartner a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1889

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-2471)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 91858 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2486-t2471-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_certauthority_root.py
Intermitten failure detected: /marvin/tests/smoke/test_primary_storage.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_routers.py
Intermitten failure detected: /marvin/tests/smoke/test_snapshots.py
Intermitten failure detected: /marvin/tests/smoke/test_vm_life_cycle.py
Intermitten failure detected: /marvin/tests/smoke/test_host_maintenance.py
Intermitten failure detected: /marvin/tests/smoke/test_hostha_kvm.py
Smoke tests completed. 61 look OK, 6 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_add_primary_storage_disabled_host | `Error` | 0.61 | test_primary_storage.py
test_01_primary_storage_nfs | `Error` | 0.13 | test_primary_storage.py
ContextSuite context=TestStorageTags>:setup | `Error` | 0.24 | test_primary_storage.py
test_04_restart_network_wo_cleanup | `Failure` | 3.27 | test_routers.py
test_02_list_snapshots_with_removed_data_store | `Error` | 1.15 | test_snapshots.py
test_08_migrate_vm | `Error` | 17.79 | test_vm_life_cycle.py
test_01_cancel_host_maintenace_with_no_migration_jobs | `Failure` | 6.34 | test_host_maintenance.py
test_02_cancel_host_maintenace_with_migration_jobs | `Error` | 2.29 | test_host_maintenance.py
test_hostha_kvm_host_fencing | `Failure` | 1491.43 | test_hostha_kvm.py
test_hostha_kvm_host_recovering | `Failure` | 616.59 | test_hostha_kvm.py

rafaelweingartner commented:
@borisstoyanov with that change in the test case, now the test we created for this PR passed. Some tests failed, but they are different from the last run.

borisstoyanov commented:
yup, let me run it again @blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

PaulAngus commented:
@blueorangutan test

blueorangutan commented:
@PaulAngus a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-2488)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 97660 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2486-t2488-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_certauthority_root.py
Intermitten failure detected: /marvin/tests/smoke/test_primary_storage.py
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermitten failure detected: /marvin/tests/smoke/test_routers.py
Intermitten failure detected: /marvin/tests/smoke/test_snapshots.py
Intermitten failure detected: /marvin/tests/smoke/test_ssvm.py
Intermitten failure detected: /marvin/tests/smoke/test_vm_life_cycle.py
Intermitten failure detected: /marvin/tests/smoke/test_host_maintenance.py
Intermitten failure detected: /marvin/tests/smoke/test_hostha_kvm.py
Smoke tests completed. 61 look OK, 6 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_add_primary_storage_disabled_host | `Error` | 0.74 | test_primary_storage.py
test_01_primary_storage_nfs | `Error` | 0.12 | test_primary_storage.py
ContextSuite context=TestStorageTags>:setup | `Error` | 0.21 | test_primary_storage.py
test_04_restart_network_wo_cleanup | `Failure` | 4.11 | test_routers.py
test_02_list_snapshots_with_removed_data_store | `Error` | 1.16 | test_snapshots.py
test_08_migrate_vm | `Error` | 18.91 | test_vm_life_cycle.py
test_01_cancel_host_maintenace_with_no_migration_jobs | `Failure` | 1.14 | test_host_maintenance.py
test_02_cancel_host_maintenace_with_migration_jobs | `Error` | 2.31 | test_host_maintenance.py
test_hostha_enable_ha_when_host_in_maintenance | `Error` | 1.56 | test_hostha_kvm.py
test_hostha_kvm_host_degraded | `Failure` | 1503.52 | test_hostha_kvm.py
test_hostha_kvm_host_fencing | `Failure` | 619.78 | test_hostha_kvm.py
test_hostha_kvm_host_recovering | `Failure` | 621.92 | test_hostha_kvm.py

rafaelweingartner commented:
@nitin-maharana are you LGTM here?

GabrielBrascher commented on server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:
```diff
@@ -2097,26 +2079,102 @@ public Volume migrateVolume(MigrateVolumeCmd cmd) {
+    }
+
+    /**
+     *  Retrieve the storage pool tags as a {@link String}. If the storage pool does not have tags we return a null value.
```
I think that  "Retrieves" would best fit here.

rafaelweingartner commented on server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:
```diff
@@ -2097,26 +2079,102 @@ public Volume migrateVolume(MigrateVolumeCmd cmd) {
-        StoragePool destPool = (StoragePool)dataStoreMgr.getDataStore(destPoolId, DataStoreRole.Primary);
-        assert (destPool != null);
+    /**
+     * Retrieve the new disk offering UUID that might be sent to replace the current one in the volume being migrated.
```
If you read as "it (the method) retrieves the new .....". Then, yes, I guess it should be "Retrieves".
I will change that.
